ports/samd/samd_spiflash.c

```diff
@@ -0,0 +1,299 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2019-2020 Peter Hinch
+ * Copyright (c) 2023 Robert Hammelrath
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+#include <stdint.h>
+#include "py/obj.h"
+#include "py/runtime.h"
+#include "py/mphal.h"
+#include "extmod/machine_spi.h"
+#include "extmod/vfs.h"
+#include "modmachine.h"
+#include "pin_af.h"
+
+#if MICROPY_HW_SPIFLASH
+
+#define _READ_INDEX (0)
+#define _PROGRAM_PAGE_INDEX (1)
+#define _SECTOR_ERASE_INDEX (2)
+
+const uint8_t _COMMANDS_24BIT[] = {0x03, 0x02, 0x20};  // READ, PROGRAM_PAGE, ERASE_4K
+const uint8_t _COMMANDS_32BIT[] = {0x13, 0x12, 0x21};  // READ, PROGRAM_PAGE, ERASE_4K
+
+#define COMMAND_JEDEC_ID (0x9F)
+#define COMMAND_READ_STATUS (0x05)
+#define COMMAND_WRITE_ENABLE (0x06)
+#define COMMAND_READ_SFDP (0x5A)
+#define PAGE_SIZE (256)
+#define SECTOR_SIZE (4096)
+
+typedef struct _spiflash_obj_t {
+    mp_obj_base_t base;
+    mp_obj_base_t *spi;
+    mp_hal_pin_obj_t cs;
+    bool addr_is_32bit;
+    uint16_t pagesize;
+    uint16_t sectorsize;
+    const uint8_t *commands;
+    uint32_t size;
+} spiflash_obj_t;
+
+extern const mp_obj_type_t spiflash_type;
```

Comment:
I think this should be named `samd_spiflash_type`, to match the other two (flash, qspi).

Reply:
Done.